### PR TITLE
Hotfix for 403 when user changed name

### PR DESCRIPTION
**Pull request by sanyavertolet**
### What's done:
 * Moved logic connected with authorization cast to `BackendService`
 * Replaced `BackendService#findByName` with `BackendService#findByAuthentication`

### CODE REVIEW

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt [hunk 1]
```diff
@@ -52,6 +55,38 @@ class BackendService(
...
+    /**
+     * Find current user [SaveUserDetails] by [authentication].
+     *
+     * @param authentication current user [Authentication]
+     * @return current user [SaveUserDetails]
+     */
+    fun findByAuthentication(authentication: Authentication): Mono<SaveUserDetails> = when (authentication) {
+        is UsernamePasswordAuthenticationToken -> findByName(authentication.name)
+        is OAuth2AuthenticationToken -> {
+            val source = authentication.authorizedClientRegistrationId
```
Comment (akuleshov7) [inline api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt:67]: I would actually move this code to some common  method...

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt [hunk 2]
```diff
@@ -52,6 +55,38 @@ class BackendService(
...
+     *
+     * @param authentication current user [Authentication]
+     * @return current username
+     */
+    fun findNameByAuthentication(authentication: Authentication?): Mono<String> = when (authentication) {
+        is UsernamePasswordAuthenticationToken -> authentication.name.toMono()
+        is OAuth2AuthenticationToken -> {
+            val source = authentication.authorizedClientRegistrationId
+            val nameInSource = authentication.name
+            findByOriginalLogin(source, nameInSource).map { it.name }
```
Comment (akuleshov7) [inline api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt:85]: and in this method just do commonMethod.name() ...

**Review by akuleshov7 — APPROVED** [review]

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt (hunk 1 above)
Comment (sanyavertolet) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt:67]: the point is that `UsernamePasswordAuthenticationToken` path varies - sometimes we want to make a request while sometimes we don't want

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt (hunk 2 above)
Comment (sanyavertolet) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt:85]: When `authentication` is `OAuth2AuthenticationToken`, we must send a request, while when `authentication` is `UsernamePasswordAuthenticationToken`, there are to different variants.

Comment (nulls) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt:85]: ok, then
```kotlin
fun findNameByAuthentication(authentication: Authentication?): Mono<String> = when (authentication) {
        is UsernamePasswordAuthenticationToken -> authentication.name.toMono()
        else -> findByAuthentication(authentication).map { it.name }
    }
```

just to minimize duplication

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt (hunk 1 above)
Comment (nulls) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt:67]: and btw, we have similar logic here `AuthorizationHeadersGatewayFilterFactory`

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt [hunk 3]
```diff
@@ -195,7 +196,7 @@ private fun userStatusBasedAuthorizationDecision(
     backendService: BackendService,
     authentication: Mono<Authentication>,
     authorizationContext: AuthorizationContext,
-) = authentication.flatMap { backendService.findByName(it.name) }
```
Comment (nulls) [inline api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt:198]: actually, you don't need to get a request to backend here -- authentication.name for basic auth is the save user name here

Comment (sanyavertolet) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/security/WebSecurityConfig.kt:198]: I do need to make a request in order to get `isActive` flag

File: api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt (hunk 2 above)
Comment (sanyavertolet) [reply api-gateway/src/main/kotlin/com/saveourtool/save/gateway/service/BackendService.kt:85]: Well, this one is ok, but there are two type casts...

**Pull request merged**